coqPackages_8_16.coq-lsp: init at 0.1.6+8.16

marsam opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Add https://github.com/ejgallego/coq-lsp

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Thesola10 commented on pkgs/development/tools/language-servers/coq-lsp/default.nix:
```diff
@@ -0,0 +1,35 @@
+
+ocamlPackages.buildDunePackage rec {
+  pname = "coq-lsp";
+  version = "0.1.3+v8.16";
```
`+v8.16` refers to the Coq version right? Should this package inherit Coq version from attributes like `coqPackages_8_16` instead?

Thesola10 commented on pkgs/development/tools/language-servers/coq-lsp/default.nix:
```diff
@@ -0,0 +1,35 @@
+
+  duneVersion = "3";
+
+  src = fetchurl {
+    url = "https://github.com/ejgallego/coq-lsp/releases/download/${version}/coq-lsp-${lib.replaceStrings [ "+" ] [ "." ] version}.tbz";
+    hash = "sha256-1tWJrhhFPZJRtCUN9Q5Zz8hyVN4ORzThPFvKBqsUzuU=";
+  };
```
We could use `fetchFromGitHub` on a specific tag (without `+v.something`) and trust that the Nix-provided version of Coq will remain constant in the versioned attr (eg. `coq_8_16`). If I understand correctly, assuming a specific version in a source bundle would break whenever the default `coq` attribute is updated.

Thesola10 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -16900,6 +16900,8 @@ with pkgs;
     llvmPackages = llvmPackages_latest;
   };
 
+  coq-lsp = callPackage ../development/tools/language-servers/coq-lsp { };
```
Should be in `coq-packages.nix` instead due to the dependency on a version of Coq.

vbgl commented:
It seems that version `0.1.4` is out, right?

marsam commented:
Thank you for reviewing. I moved the expression to `coq-packages.nix` and updated to 0.1.4. 
Tomorrow I will test if it's still working, I'm a bit busy with work.

Thesola10 reviewed (approved):
LGTM

vbgl commented on pkgs/top-level/coq-packages.nix:
```diff
@@ -33,6 +33,7 @@ let
       coq-bits = callPackage ../development/coq-modules/coq-bits {};
       coq-elpi = callPackage ../development/coq-modules/coq-elpi {};
       coq-ext-lib = callPackage ../development/coq-modules/coq-ext-lib {};
+      coq-lsp = callPackage ../development/coq-modules/coq-lsp {};
       coq-haskell = callPackage ../development/coq-modules/coq-haskell { };
```
Can we try to keep this list in alphabetical order?

vbgl commented on pkgs/development/coq-modules/coq-lsp/default.nix:
```diff
@@ -0,0 +1,37 @@
+    { case = range "8.16" "8.17"; out = "0.1.4+v8.16"; }
+  ] null;
+
+  buildPhase = ''
+    runHook preBuild
+    dune build -p ${pname} -j $NIX_BUILD_CORES
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    dune install --prefix $out ${pname}
+    runHook postInstall
+  '';
+
```
Isn’t that very close to the default values of `buildPhase` and `installPhase`? If possible I’d prefer to avoid redefining them.

marsam replied:
I updated the expression to use the default `buildPhase`, but I can't do the same with `installPhase` because it tries to install `$out/lib/coq` which is not present in coq-lsp
https://github.com/NixOS/nixpkgs/blob/be1950af0b7732ea7ef53634ef94a64956424ed3/pkgs/build-support/coq/default.nix#L119-L122

ejgallego commented on pkgs/development/coq-modules/coq-lsp/default.nix:
```diff
@@ -0,0 +1,37 @@
+
+  release."0.1.4+v8.16".sha256 = "sha256-5jztvAKpCFehtxu9Kw2inbPUZYVFIMcIoD5/u1WBuyI=";
+
+  inherit version;
+  defaultVersion = with lib.versions; lib.switch coq.coq-version [
+    { case = range "8.16" "8.17"; out = "0.1.4+v8.16"; }
+  ] null;
```
Folks thanks for this PR, for Coq 8.17 you need a 8.17 specific tree, unfortunately the Coq API we use is not stable between versions.

Tag for 8.17 is 0.1.4+v8.17

Thesola10 replied:
Then it should be possible to use `"0.1.4+v${coq.version}"`, that's the beauty about a functional language

ejgallego replied:
It should yes! Something that has been brought to my attention is whether the branch names should be `+v8.16` or `+8.16`, it seems the lastest scheme is the standard (including in some of my packages) so I'm unsure if I should tweak the tag names for the upcoming releases (and add the corresponding aliases back)

Any advice you folks have here as to easy packaging is most welcome.

Thesola10 replied:
It would be preferable, as I think the v messes up the reverse package name parser (e.g. when doing `nix-env -e coq-lsp` it tries to strip the version suffix)

ejgallego replied:
Thanks @Thesola10 , so to confirm `0.1.4+8.16` is Nix-friendly ?

Thesola10 replied:
You might not need to change your actual tagging scheme -- since `mkDerivation` accepts arbitrary key/values, you could compose the derivation's `version` as `"${packageVersion}+${coq.coq-version}"` and the source tag as `"${packageVersion}+v${coq.coq-version}"`

Alizter commented:
Hi, coq-lsp 0.1.5 has been released. Would we be able to bump the version here?

Alizter commented:
Also we have a flake capable of building that has been merged, does that help things?

marsam committed:
coqPackages_8_16.serapi: 8.16.0+0.16.0 -> 8.16.0+0.16.3

marsam committed:
coqPackages_8_16.coq-lsp: init at 0.1.6.1+8.16

ejgallego commented:
Hi @Alizter I'd suggest packaging 0.1.6 which is just out.

coq-lsp is a standard dune package (an vscode extension) so as long as Coq kinda works things should be fine, it also obeys `COQLIB` and `COQPATH` if Nix is using them.

marsam commented:
Thank you for your reviews. I updated the expression to `0.1.6+8.16` and tested it works correctly.

vbgl reviewed (approved):
I did not test, but it looks good.

ejgallego commented:
Great @marsam , thank you very much!

We've detected a bug specific to 8.16 support (we usually test in 8.17), it is a small bug but annoying, we put a new https://github.com/ejgallego/coq-lsp/releases/tag/0.1.6.1%2B8.16 tag with the 8.16 specific fix.

marsam commented:
Updated to `0.1.6.1+8.16`
